src/shim/binary_spinning_sem.cc

```diff
@@ -42,3 +44,18 @@ void BinarySpinningSem::wait(bool spin) {
     _x.store(false, std::memory_order_release);
     pthread_spin_unlock(&_lock);
 }
+
+int BinarySpinningSem::trywait() {
+    int rv = sem_trywait(&_semaphore);
+    if (rv != 0) {
+        assert(rv == -1 && errno == EAGAIN);
```

Comment:
Any chance for us to receive an interrupt/signal here? Maybe we should allow for `EINTR`